x-pack/plugins/fleet/server/services/epm/packages/install.ts

```diff
@@ -291,20 +292,43 @@ export async function handleInstallPackageFailure({
       status: 'install_failed',
       latestInstallFailedAttempts,
     });
-
-    if (installType === 'reinstall') {
-      logger.error(`Failed to reinstall ${pkgkey}: [${error.toString()}]`, { error });
+    // in case of install, uninstall any package assets
+    if (installType === 'install') {
+      logger.error(
+        `Uninstalling ${pkgkey} after error installing: [${error.toString()}] with install type: ${installType}`
+      );
+      await removeInstallation({ savedObjectsClient, pkgName, pkgVersion, esClient });
+      return;
     }
 
+    // in case of reinstall, restart install where it left off
+    // retry MAX_REINSTALL_RETRIES times before exiting, in case the error persists
+    if (installType === 'reinstall' && attemptNumber < MAX_REINSTALL_RETRIES) {
+      logger.error(`Error installing ${pkgkey}: [${error.toString()}]`);
+      logger.debug(
+        `Retrying install of ${pkgkey} with install type: ${installType} - Attempt ${attemptNumber} `
+      );
+      await installPackage({
+        installSource: 'registry',
+        savedObjectsClient,
+        pkgkey,
+        esClient,
+        spaceId,
+        authorizationHeader,
+        retryFromLastState: true,
```

Comment:
I think the `retryFromLastState` should be used even when `MAX_REINSTALL_RETRIES` is reached.
Now if the max retries is reached once, and the package is reinstalled without a force flag, the reinstall will happen from the first step, not the last successful step.

Reply:
The reason for adding `MAX_REINSTALL_RETRIES` is that, in case of persistent error, we get into an infinite loop. The code flow is very convoluted but when I tested locally I realised there was no exit condition for a case like this, so I added the retries number. 
It's probably very rare that an user incurs in such a case, but I'd like to avoid such a situation. 

> Now if the max retries is reached in a previous reinstall, and the package is reinstalled without a force flag, the reinstall will happen from the first step, not the last successful step.
I think it's ok to restart from the first state in this case, if the reinstall happened already for some retries and didn't succees, we probably want to reinstall from scratch instead. What do you think?

Another thing could be to increase the number of retries to a bigger number, for instance 5 instead of 3.

Reply:
> I think it's ok to restart from the first state in this case, if the reinstall happened already for some retries and didn't succees, we probably want to reinstall from scratch instead. What do you think?

Yeah it probably indicates a persistent problem if all retries failed. I'm fine with restarting from the first step then.